pyomo/core/base/units_container.py

```diff
@@ -0,0 +1,1335 @@
+        expr.NPV_ProductExpression: _get_unit_for_product,
+        expr.ReciprocalExpression: _get_unit_for_reciprocal,
+        expr.NPV_ReciprocalExpression: _get_unit_for_reciprocal,
+        expr.PowExpression: _get_unit_for_pow,
+        expr.NPV_PowExpression: _get_unit_for_pow,
+        expr.NegationExpression: _get_unit_for_single_child,
+        expr.NPV_NegationExpression: _get_unit_for_single_child,
+        expr.AbsExpression: _get_unit_for_single_child,
+        expr.NPV_AbsExpression: _get_unit_for_single_child,
+        expr.UnaryFunctionExpression: _get_unit_for_unary_function,
+        expr.NPV_UnaryFunctionExpression: _get_unit_for_unary_function,
+        expr.Expr_ifExpression: _get_unit_for_expr_if,
+        IndexTemplate: _get_dimensionless_no_children,
+        expr.GetItemExpression: _get_dimensionless_with_dimensionless_children,
+        expr.ExternalFunctionExpression: _get_dimensionless_with_dimensionless_children,
+        expr.NPV_ExternalFunctionExpression: _get_dimensionless_with_dimensionless_children,
+        expr.LinearExpression: _get_unit_for_linear_expression
+    }
+
+    unary_function_method_map = {
+        'log': _get_dimensionless_with_dimensionless_children,
+        'log10':_get_dimensionless_with_dimensionless_children,
+        'sin': _get_dimensionless_with_radians_child,
+        'cos': _get_dimensionless_with_radians_child,
+        'tan': _get_dimensionless_with_radians_child,
+        'sinh': _get_dimensionless_with_radians_child,
+        'cosh': _get_dimensionless_with_radians_child,
+        'tanh': _get_dimensionless_with_radians_child,
+        'asin': _get_radians_with_dimensionless_child,
+        'acos': _get_radians_with_dimensionless_child,
+        'atan': _get_radians_with_dimensionless_child,
+        'exp': _get_dimensionless_with_dimensionless_children,
+        'sqrt': _get_unit_sqrt,
+        'asinh': _get_radians_with_dimensionless_child,
+        'acosh': _get_radians_with_dimensionless_child,
+        'atanh': _get_radians_with_dimensionless_child,
+        'ceil': _get_unit_for_single_child,
+        'floor': _get_unit_for_single_child
+    }
+
+    def exitNode(self, node, data):
+        """ Callback for :class:`pyomo.core.current.StreamBasedExpressionVisitor`. This
+        method is called when moving back up the tree in a depth first search."""
+        if _is_leaf(node):
+            if isinstance(node, _PyomoUnit):
+                return (node, node._get_pint_unit())
+
+            # ToDo: Check for Var or Param and return their units...
+            # I have a leaf, but this is not a PyomoUnit - (treat as dimensionless)
+            return (None, None)
+
+        # not a leaf - get the appropriate function for type of the node
+        node_type = type(node)
+        if node_type in self.node_type_method_map:
+            node_func = self.node_type_method_map[node_type]
+            if node_func is not None:
```

Comment:
This seems more complicated than it needs to be.  Why not:
```python
node_func = self.node_type_method_map.get(type(node), None)
if node_func is not None:
    #...
```

Reply:
Good call. I frequently forget about the .get with a default.